skbeam/core/correlation.py

```diff
@@ -405,3 +406,357 @@ def auto_corr_scat_factor(lags, beta, relaxation_rate, baseline=1):
+                                                 fi_binned))*num_pixels
+        else:
+            two_time[tind1, tind2] = tmp_binned/(pi_binned *
+                                                   fi_binned)*num_pixels
+
+
+def _init_state_two_time(num_levels, num_bufs, labels, num_frames):
+    """Initialize a stateful namedtuple for the multi-tau
+     for two time correlation
+
+    Parameters
+    ----------
+    num_levels : int
+    num_bufs : int
+    labels : array
+        Two dimensional labeled array that contains ROI information
+
+    Returns
+    -------
+    internal_state : namedtuple
+        The namedtuple that contains all the state information that
+        `lazy_one_time` requires so that it can be used to pick up processing
+        after it was interrupted
+    """
+    (label_array, pixel_list, num_rois, num_pixels, lag_steps, buf,
+     img_per_level, track_level, cur) = _inputs(num_bufs, num_levels, labels)
+
+    # to count images in each level
+    count_level = np.zeros(num_levels, dtype=np.int64)
+    # current image time
+    #current_img_time = 0
+
+    # two time correlation results (array)
+    two_time = np.zeros((num_frames, num_frames,
+                         num_rois), dtype=np.float64)
+
+    return _two_time_internal_state(
+        buf,
+        img_per_level,
+        label_array,
+        track_level,
+        cur,
+        pixel_list,
+        num_pixels,
+        lag_steps,
+        two_time,
+        count_level,
+        #current_img_time,
+    )
+
+
+def _inputs(num_bufs, num_levels, labels):
+    """
+    This is a helper function to validate inputs and create initial state
+    inputs for both one time and two time correlation
+
+    Parameters
+    ----------
+    num_bufs : int
+    num_levels : int
+    labels : array
+        labeled array of the same shape as the image stack;
+        each ROI is represented by a distinct label (i.e., integer)
+
+    Returns
+    -------
+    label_array : array
+        labels of the required region of interests(ROI's)
+    pixel_list : array
+        1D array of indices into the raveled image for all
+        foreground pixels (labeled nonzero)
+        e.g., [5, 6, 7, 8, 14, 15, 21, 22]
+    num_rois : int
+        number of region of interests (ROI)
+    num_pixels : array
+        number of pixels in each ROI
+    lag_steps : array
+        the times at which the correlation was computed
+    buf : array
+        image data for correlation
+    img_per_level : array
+        to track how many images processed in each level
+    track_level : array
+        to track processing each level
+    cur : array
+
+    """
+    if num_bufs % 2 != 0:
+        raise ValueError("There must be an even number of `num_bufs`. You "
+                         "provided %s" % num_bufs)
+    label_array, pixel_list = extract_label_indices(labels)
+
+    # map the indices onto a sequential list of integers starting at 1
```

Comment:
Err, sorry, this is supposed to be on the line below this comment.

To elaborate a little... It seems that by mapping these onto a list starting at 1 (instead of at 0), it means that every time you do a `np.bincount`, you have to add a `[1:]` to the end of the resulting array.